src/consensus/dkg.rs

```diff
@@ -0,0 +1,223 @@
+        )
+    }
+}
+
+/// Returns the number of minumn responsive participants expected for the DKG process
+#[inline]
+pub const fn threshold_count(elder_size: usize) -> usize {
+    // TODO: allow threshold to be configurable.
+    elder_size * QUORUM_NUMERATOR / QUORUM_DENOMINATOR
+}
+
+pub type DkgKey = (BTreeSet<PublicId>, u64);
+
+#[derive(Default)]
+pub struct DkgVoter {
+    dkg_cache: BTreeMap<DkgKey, EldersInfo>,
+    key_gen_map: BTreeMap<DkgKey, KeyGen<FullId>>,
+    pending_accumulated_events: VecDeque<(AccumulatingEvent, Proof)>,
+    completed_dkg: BTreeSet<DkgKey>,
+    sibling_dkg_result_cache: BTreeMap<DkgKey, DkgResult>,
+    timer_token: u64,
+}
+
+impl DkgVoter {
+    pub fn check_dkg(
+        &mut self,
+    ) -> (
+        BTreeMap<DkgKey, DkgResult>,
+        VecDeque<(AccumulatingEvent, Proof)>,
+    ) {
+        let mut completed = BTreeMap::new();
+        for (key, key_gen) in self.key_gen_map.iter_mut() {
+            if key_gen.is_finalized() {
+                let dkg_key = (key.0.clone(), key.1);
+                if let Some((_participants, dkg_outcome)) = key_gen.generate_keys() {
+                    let dkg_result = DkgResult::new(
+                        dkg_outcome.public_key_set,
+                        Some(dkg_outcome.secret_key_share),
+                    );
+                    let _ = completed.insert(dkg_key.clone(), dkg_result);
+                }
+                let _ = self.completed_dkg.insert(dkg_key);
+            }
+        }
+
+        // Only handle cached accumulated events after DKG completion, if has.
+        let backlog_events = if !completed.is_empty() {
+            mem::replace(&mut self.pending_accumulated_events, VecDeque::new())
+        } else {
+            VecDeque::new()
+        };
+
+        (completed, backlog_events)
+    }
+
+    pub fn remove_voter(&mut self, dkg_key: &DkgKey) {
+        let _ = self.key_gen_map.remove(dkg_key);
+    }
+
+    pub fn progress_dkg(&mut self, rng: &mut MainRng) -> Vec<(DkgKey, DkgMessage<PublicId>)> {
+        let mut broadcast = Vec::new();
+        for (key, key_gen) in self.key_gen_map.iter_mut() {
+            debug!("Progressing DKG {:?}", key);
+            if let Ok(messages) = key_gen.timed_phase_transition(rng) {
+                for message in messages {
+                    broadcast.push((key.clone(), message));
+                }
+            }
+        }
+
+        broadcast
+    }
+
+    pub fn process_dkg_message(
+        &mut self,
+        rng: &mut MainRng,
+        dkg_key: &DkgKey,
+        message: DkgMessage<PublicId>,
+    ) -> Vec<DkgMessage<PublicId>> {
+        let mut messages = Vec::new();
+
+        if let Some(mut key_gen) = self.key_gen_map.remove(dkg_key) {
+            if let Ok(responses) = key_gen.handle_message(rng, message) {
+                messages = responses;
+            }
+
+            let _ = self.key_gen_map.insert(dkg_key.clone(), key_gen);
+        }
```

Comment:
Why are you removing and then reinserting it here? Can't you do just:
```rust
        if let Some(key_gen) = self.key_gen_map.get_mut(dkg_key) {
            if let Ok(responses) = key_gen.handle_message(rng, message) {
                messages = responses;
            }
        }
```

Reply:
Ah, previously there is extra work be done by `approved`, which causing borrow check error.
with the squashing, it shall now be ok.
I will change it.